File: Extras/Gaia.Ajax.GPL/4.0 - 31.10.2013/Source/Gaia.WebWidgets.CommunityExtensions/Scripts/FishEyeMenu.js
// Gaia Ajax Copyright (C) 2008 - 2013 Gaiaware AS. details at http://gaiaware.net/

/* 
 * Gaia Ajax - Ajax Control Library for ASP.NET
 * Copyright (C) 2008 - 2013 Gaiaware AS
 * All rights reserved.
 * This program is distributed under either GPL version 3 
 * as published by the Free Software Foundation or the
 * Gaia Commercial License version 1 as published by
 * Gaiaware AS
 * read the details at http://gaiaware.net
 */

/* ---------------------------------------------------------------------------
   Class wrapping a FishEyeMenu that fakes the Mac OS X taskbar
   --------------------------------------------------------------------------- */

if( !Gaia.Extensions )
  Gaia.Extensions = Class.create();
 
Gaia.Extensions.FishEyeMenu = jsface.Class(Gaia.Panel, {

  constructor: function (element, options) {
    Gaia.Extensions.FishEyeMenu.$super.call(this, element, options);
    this.initializeFishEyeMenu(element, options);
  },

  initializeFishEyeMenu: function(element, options) {

    // set properties 
    this.options.startSize = this.options.startSize || 55;
    this.options.endSize = this.options.endSize || 88;
    this.options.threshold = this.options.threshold || 300;
    this._currentlyEnabled = this.options.enabled || true;


    // all image elements within this element.
    this.animateElements = $$('#' + this.element[0].id + ' input[type="image"]');
    var tmpArray = $$('#' + this.element[0].id + ' img')
    tmpArray.each(function(idx){
      this.animateElements.push(idx);
    }.bind(this));
    Element.observe(document, 'mousemove', this.onBodyMove.bindAsEventListener(this));
  },
  
  setEnabled: function(value) {
    this._currentlyEnabled = value;
    return this;
  },

  isEnabled: function() {
    return this._currentlyEnabled;
  },

  onBodyMove: function(event) {
    if (!this._currentlyEnabled) {
      return;
    }
    
    for (var y = 0; y < this.animateElements.length; y ++) {
      var s = this.animateElements[y];
      var position = Position.cumulativeOffset(s);
      var xOff = Math.abs(Event.pointerX(event) - (position[0] + (s.offsetWidth / 2)));
      var yOff = Math.abs(Event.pointerY(event) - (position[1] + (s.offsetHeight / 2)));
      var offset = Math.sqrt((xOff*xOff)+(yOff*yOff));
      if( offset < this.options.threshold ){
        offset = offset || 1; // Prevent divide by zero bugs
        var delta = this.options.endSize - this.options.startSize;
        var factorIncrease = 1 - (offset / this.options.threshold);
        s.style.width = (parseInt(this.options.startSize + (delta * factorIncrease), 10)) + 'px';
        s.style.height = (parseInt(this.options.startSize + (delta * factorIncrease), 10)) + 'px';
      } else {
        if( parseInt(s.style.width, 10) != this.options.startSize ){
          s.style.width = this.options.startSize+'px';
          s.style.height = this.options.startSize+'px';
        }
      }
    }
  }
});

File: Extras/Gaia.Ajax.GPL/4.0 - 31.10.2013/Source/Gaia.WebWidgets.CommunityExtensions/Widgets/FishEyeMenu.cs
/*******************************************************************
 * Gaia Ajax - Ajax Control Library for ASP.NET  
 * Copyright (C) 2008 - 2011 Gaiaware AS
 * All rights reserved. 
 * This program is distributed under either GPL version 3 
 * as published by the Free Software Foundation or the
 * Gaia Commercial License version 1 as published by Gaiaware AS
 * read the details at http://gaiaware.net/product/dual-licensing 
 ******************************************************************/

using System.ComponentModel;
using System.Web.UI;
using System.Drawing;

namespace Gaia.WebWidgets.CommunityExtensions
{
    /// <summary>
    /// The Gaia Ajax FishEyeMenu is like the dock menu of Mac OS X.
    /// The Gaia Ajax FishEyeMenu is just a normal panel that will find all image elements within it during initialization 
    /// and add up logic to zoom into them as you hover over them. To use it just add up any widget (or HTML elements) that 
    /// renders "img" elements and set your properties. You can easily combine it together with a Repeater, add up 
    /// ImageButtons (if you want to trap clicks of elements in it). You can create Ajax FishEye Menus containing anything you want.
    /// </summary>
    [DefaultProperty("CssClass")]
    [ToolboxData("<{0}:FishEyeMenu runat=\"server\"></{0}:FishEyeMenu>")]
    [ToolboxBitmap(typeof(FishEyeMenu), "Resources.Gaia.WebWidgets.CommunityExtensions.FishEye.bmp")]
    public class FishEyeMenu : Panel, IAjaxControl
    {
        #region [ -- Properties -- ]
        /// <summary>
        /// What size the images are at default when not hovered over
        /// </summary>
        [DefaultValue(55)]
        public int StartSize
        {
            get
            {
                if (ViewState["StartSize"] == null)
                    return 55;
                return (int)ViewState["StartSize"];   
            }
            set
            {
                ViewState["StartSize"] = value;
            }
        }

        /// <summary>
        /// Maximum size the images will resize to
        /// </summary>
        [DefaultValue(88)]
        public int EndSize
        {
            get
            {
                if (ViewState["EndSize"] == null)
                    return 88;
                return (int)ViewState["EndSize"];
            }
            set
            {
                ViewState["EndSize"] = value;
            }
        }

        /// <summary>
        /// At what distance the images will start blowing up in size
        /// </summary>
        [DefaultValue(300)]
        public int Threshold
        {
            get
            {
                if (ViewState["Threshold"] == null)
                    return 300;
                return (int)ViewState["Threshold"];
            }
            set
            {
                ViewState["Threshold"] = value;
            }
        }
        #endregion

        #region [ -- Overridden Base class methods -- ]

        /// <summary>
        /// Include FishEyeMenu Javascript files
        /// </summary>
        protected override void IncludeScriptFiles()
        {
            base.IncludeScriptFiles();

            // Include FishEyeMenu Javascript stuff
            Manager.Instance.AddInclusionOfFileFromResource("Gaia.WebWidgets.CommunityExtensions.Scripts.prototype.js", typeof(FishEyeMenu), "");
            Manager.Instance.AddInclusionOfFileFromResource("Gaia.WebWidgets.CommunityExtensions.Scripts.FishEyeMenu.js", typeof(FishEyeMenu), "Gaia.Extensions.FishEyeMenu.browserFinishedLoading");
        }

        #endregion

        #region [ -- Overridden IAjaxControl methods -- ]

        string IAjaxControl.GetScript()
        {
            return new RegisterControl("Gaia.Extensions.FishEyeMenu", ClientID)
                .AddProperty("startSize", StartSize, false)
                .AddProperty("endSize", EndSize, false)
                .AddProperty("threshold", Threshold, false)
                .AddProperty("enabled", Enabled)
                .AddAspects(Aspects).ToString();
        }

        #endregion
    }
}
